File: swig/R/test_linalg.R
test_Sort <- function() {
   x = rnorm(2000000,0,1)
   return(Xtest(quote(sort(x)),quote(spams.Sort(x,TRUE))))
}

test_CalcAAt <- function () {
  m = 200;n = 200000; d= 0.05
  A = rSpMatrix(m,n,floor(m * n * d))
  return(Xtest(quote(A %*% t(A)),quote(spams.CalcAAt(A))))
}

test_CalcXAt <- function () {
  m = 200;n = 200000; d= 0.05
  A = rSpMatrix(m,n,floor(m * n * d))
  X = matrix(rnorm(64 * n),nrow = 64,ncol = n)
  return(Xtest(quote(X %*% t(A)),quote(spams.CalcXAt(X,A))))
}
matprod <- function(x,y) "%*%"(x,y)
test_CalcXY <- function() {
    X = matrix(rnorm(64 * 200),nrow = 64,ncol = 200,byrow = FALSE)
    Y = matrix(rnorm(200 * 20000),nrow = 200,ncol = 20000,byrow = FALSE)
    return(Xtest(quote(X %*% Y),quote(spams.CalcXY(X,Y))))
}

test_CalcXYt <- function () {
    X = matrix(rnorm(64 * 200),nrow = 64,ncol = 200,byrow = FALSE)
    Y = matrix(rnorm(200 * 20000),nrow = 20000,ncol = 200,byrow = FALSE)
    return(Xtest(quote(X %*% t(Y)),quote(spams.CalcXYt(X,Y))))
  
}

test_CalcXtY <- function () {
    X = matrix(rnorm(64 * 200),nrow = 200,ncol = 64,byrow = FALSE)
    Y = matrix(rnorm(200 * 20000),nrow = 200,ncol = 20000,byrow = FALSE)
    return(Xtest(quote(t(X) %*% Y),quote(spams.CalcXtY(X,Y))))
}

test_Bayer <- function () {
  X = rnorm(2000000,0,1)
  y2 = Xtest1('spams',quote(spams.Bayer(X,0)),n=1)
  return(NULL)
}

test_ConjGrad <- function () {
  A = matrix(rnorm(500 * 5000),nrow = 5000,ncol = 500)
  A = t(A) %*% A
  b = as.vector(rep(1,ncol(A)))
  x0 = b
  tol = 1e-4
  itermax = floor(0.5 * length(b))
  CGtest <- function(txt,expr) {
    tic = proc.time()
    for (i in 1:20)
      y = eval(expr)
    tac = proc.time()
    cat(sprintf("  Time (%s): %f\n",txt,(tac - tic)[['user.self']]))
    x1 = abs(b - (A %*% y))
    cat(sprintf("Mean error on b : %f\n\n",sum(x1) / length(b)))
    return(y)
  }
  y2 = CGtest("R",quote(solve(A,b)))
  y1 = CGtest("spams",quote(spams.ConjGrad(A,b,x0,tol,itermax)))
  ## y1 = CGtest("spams",quote(spams.ConjGrad(A,b)))
  return(max(abs(y1 - y2)))
}

test_InvSym <- function() {
    A = matrix(runif(1000 * 1000,0,1),nrow = 1000,ncol = 1000,byrow = FALSE)
    A = t(A) %*% A
    return(Xtest(quote(solve(A)),quote(spams.InvSym(A))))
}
test_Normalize <- function() {
    A = matrix(runif(100 * 1000,0,1),nrow = 100,ncol = 1000)
    y2 = Xtest1("spams",quote(spams.Normalize(A)),n=1)
    return(NULL)
}

test_linalg.tests =list( 'Sort' = test_Sort,
  'CalcAAt' = test_CalcAAt,
  'CalcXAt' = test_CalcXAt,
  'CalcXY' = test_CalcXY,
  'CalcXYt' = test_CalcXYt,
  'CalcXtY' = test_CalcXtY,
  'Bayer' = test_Bayer,
  'ConjGrad' = test_ConjGrad,
  'InvSym' = test_InvSym,
  'Normalize' = test_Normalize
  )

File: swig/python/test_linalg.py
import sys
import numpy as np
import scipy
import scipy.sparse as ssp
import spams
import time
from test_utils import *

if not ('rand' in ssp.__dict__):
    import myscipy_rand
    ssprand = myscipy_rand.rand
else:
    ssprand = ssp.rand

def test_Sort():
    n = 2000000
    X = np.random.normal(size = (n,))
    return Xtest('np.sort(X)','spams.Sort(X,True)',locals())


def test_CalcAAt():
    """
    test A * A'
    """
    m=200; n = 200000; d= 0.05
    A = ssprand(m,n,density=d,format='csc',dtype=np.float64)
    return Xtest('np.dot(A,A.T)','spams.CalcAAt(A)',locals())

def test_CalcXAt():
    m=200; n = 200000; d= 0.05
    A = ssprand(m,n,density=d,format='csc',dtype=np.float64)
    X = np.asfortranarray(np.random.normal(size = (64,n)))

    # dot is very very slow betewwen a full and a sparse matrix
    return Xtest('np.dot(X,A.T.todense())','spams.CalcXAt(X,A)',locals())

def test_CalcXY():
    X = np.asfortranarray(np.random.normal(size = (64,200)))
    Y = np.asfortranarray(np.random.normal(size = (200,20000)))
    return Xtest('np.dot(X,Y)','spams.CalcXY(X,Y)',locals())

def test_CalcXYt():
    X = np.asfortranarray(np.random.normal(size = (64,200)))
    Y = np.asfortranarray(np.random.normal(size = (20000,200)))
    return Xtest('np.dot(X,Y.T)','spams.CalcXYt(X,Y)',locals())

def test_CalcXtY():
    X = np.asfortranarray(np.random.normal(size = (200,64)))
    Y = np.asfortranarray(np.random.normal(size = (200,20000)))
    return Xtest('np.dot(X.T,Y)','spams.CalcXtY(X,Y)',locals())

def test_Bayer():
    n = 2000000
    X = np.random.normal(size = (n,))

    Z = Xtest1('spams','spams.Bayer(X,0)',locals())
    return None

def test_ConjGrad():
    A = np.asfortranarray(np.random.normal(size = (5000,500)))
#    np.random.seed(0)
#    A = np.asfortranarray(np.random.normal(size = (10,5)))
    A = np.asfortranarray(np.dot(A.T,A))
    b = np.ones((A.shape[1],),dtype=np.float64,order="FORTRAN")
    x0 = b
    tol = 1e-4
    itermax = int(0.5 * len(b))

    tic = time.time()
    for i in xrange(0,20):
        y1 = np.linalg.solve(A,b)
    tac = time.time()
    print "  Time (numpy): ", tac - tic
    x1 = np.abs(b - np.dot(A,y1))
    print "Mean error on b : %f" %(x1.sum() / b.shape[0])

    tic = time.time()
    for i in xrange(0,20):
        y2 = spams.ConjGrad(A,b,x0,tol,itermax)
#        y2 = spams.ConjGrad(A,b)
    tac = time.time()
    print "  Time (spams): ", tac - tic
    x1 = np.dot(A,y2)
    x2 = np.abs(b - x1)
    print "Mean error on b : %f" %(x2.sum() / b.shape[0])

    err = abs(y1 - y2)
    return err.max()

def test_InvSym():
    A = np.asfortranarray(np.random.random(size = (1000,1000)))
    A =np.asfortranarray( np.dot(A.T,A))
    return Xtest('np.linalg.inv(A)','spams.InvSym(A)',locals())

def test_Normalize():
    A = np.asfortranarray(np.random.random(size = (100,1000)))
    res2 = Xtest1('spams','spams.Normalize(A)',locals())
    return None

tests = {
    'Sort' : test_Sort,
    'CalcAAt' : test_CalcAAt,
    'CalcXAt' : test_CalcXAt,
    'CalcXY' : test_CalcXY,
    'CalcXYt' : test_CalcXYt,
    'CalcXtY' : test_CalcXtY,
    'Bayer' : test_Bayer,
    'ConjGrad' : test_ConjGrad,
    'InvSym' : test_InvSym,
    'Normalize' : test_Normalize,
    }
